File: frontend/js/schedules.js
// Train Scheduling Management & AI Recommendations JS

let allSchedulesList = [];
let activeScheduleLineFilter = 'All';

async function loadSchedulesPage() {
    try {
        const res = await fetch(`${API_BASE}/schedules`);
        const data = await res.json();

        if (data.schedules) {
            allSchedulesList = data.schedules;
            filterSchedulesByLine(activeScheduleLineFilter);
        }

        loadAIRecommendations();
    } catch (err) {
        console.error("Error loading schedules:", err);
        showToast("Failed to load train schedules.", "error");
    }
}

function filterSchedulesByLine(line) {
    activeScheduleLineFilter = line;
    document.querySelectorAll('.line-sched-filter').forEach(btn => btn.classList.remove('active'));
    const activeBtn = document.getElementById(`sched-filter-${line}`);
    if (activeBtn) activeBtn.classList.add('active');

    if (line === 'All') {
        renderSchedulesTable(allSchedulesList);
    } else {
        const filtered = allSchedulesList.filter(s => s.line === line);
        renderSchedulesTable(filtered);
    }
}

function renderSchedulesTable(schedules) {
    const tbody = document.getElementById('schedules-tbody');
    if (!tbody) return;

    if (schedules.length === 0) {
        tbody.innerHTML = `<tr><td colspan="8" style="text-align:center; color:var(--text-muted); padding:2rem;">No schedules found for this line.</td></tr>`;
        return;
    }

    const isAdmin = currentUser && currentUser.role === 'admin';
    const isOperator = currentUser && (currentUser.role === 'operator' || currentUser.role === 'admin');

    tbody.innerHTML = schedules.map(s => {
        let statusBadgeClass = 'status-low';
        if (s.status === 'Delayed') statusBadgeClass = 'status-high';
        if (s.status === 'Cancelled') statusBadgeClass = 'status-critical';
        if (s.status.includes('Recommended')) statusBadgeClass = 'status-moderate';

        return `
            <tr>
                <td><strong>${s.train_code}</strong></td>
                <td><span style="color:var(--accent-cyan); font-weight:600;">${s.line}</span></td>
                <td>${s.origin_station} ➔ ${s.destination_station}</td>
                <td>${s.departure_time}</td>
                <td>${s.arrival_time}</td>
                <td><strong>Every ${s.frequency_mins} mins</strong></td>
                <td><span class="status-badge ${statusBadgeClass}">${s.status}</span></td>
                <td>
                    ${isOperator ? `
                        <select onchange="updateScheduleStatus(${s.id}, this.value)" class="form-select" style="padding:0.25rem 0.5rem; font-size:0.8rem; width:auto; display:inline-block;">
                            <option value="On Time" ${s.status === 'On Time' ? 'selected' : ''}>On Time</option>
                            <option value="Delayed" ${s.status === 'Delayed' ? 'selected' : ''}>Delayed</option>
                            <option value="Cancelled" ${s.status === 'Cancelled' ? 'selected' : ''}>Cancelled</option>
                        </select>
                    ` : ''}
                    ${isAdmin ? `
                        <button onclick="deleteSchedule(${s.id})" class="btn btn-danger btn-sm" style="padding:0.2rem 0.5rem; margin-left:0.25rem;">Delete</button>
                    ` : ''}
                </td>
            </tr>
        `;
    }).join('');
}

async function updateScheduleStatus(scheduleId, newStatus) {
    try {
        const res = await fetch(`${API_BASE}/schedules/${scheduleId}`, {
            method: 'PUT',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ status: newStatus })
        });
        if (res.ok) {
            showToast("Train status updated successfully.", "success");
            loadSchedulesPage();
        } else {
            showToast("Failed to update schedule status.", "error");
        }
    } catch (err) {
        showToast("Network error updating schedule.", "error");
    }
}

async function deleteSchedule(scheduleId) {
    if (!confirm("Are you sure you want to delete this train schedule?")) return;

    try {
        const res = await fetch(`${API_BASE}/schedules/${scheduleId}`, { method: 'DELETE' });
        if (res.ok) {
            showToast("Train schedule deleted.", "info");
            loadSchedulesPage();
        } else {
            showToast("Failed to delete schedule.", "error");
        }
    } catch (err) {
        showToast("Error deleting schedule.", "error");
    }
}

// AI Dispatch & Frequency Recommendation Engine
async function loadAIRecommendations() {
    const container = document.getElementById('ai-recommendations-container');
    if (!container) return;

    try {
        const res = await fetch(`${API_BASE}/schedules/recommendations`);
        const data = await res.json();

        if (data.recommendations) {
            container.innerHTML = data.recommendations.map(rec => `
                <div class="card" style="margin-bottom: 1rem; border-left: 4px solid ${rec.recommended_frequency <= 5 ? 'var(--accent-amber)' : 'var(--accent-emerald)'};">
                    <div style="display: flex; justify-content: space-between; align-items: flex-start; margin-bottom: 0.5rem;">
                        <div>
                            <strong style="font-size: 1.1rem; color: var(--accent-cyan);">${rec.line} Optimization</strong>
                            <span style="font-size: 0.85rem; color: var(--text-muted); display: block; margin-top: 0.2rem;">Busiest Hub: <strong>${rec.busiest_station}</strong> (${rec.current_max_density}% density)</span>
                        </div>
                        <span class="status-badge ${rec.recommended_frequency <= 5 ? 'status-high' : 'status-low'}">${rec.action}</span>
                    </div>
                    <p style="font-size: 0.9rem; color: var(--text-main); margin-bottom: 0.75rem;">💡 ${rec.reason}</p>
                    <div style="display: flex; gap: 1.5rem; font-size: 0.85rem; background: rgba(0,0,0,0.25); padding: 0.6rem 1rem; border-radius: var(--radius-sm);">
                        <div>Current Frequency: <strong>${rec.current_avg_frequency} mins</strong></div>
                        <div>AI Recommended Frequency: <strong style="color: var(--accent-cyan);">${rec.recommended_frequency} mins</strong></div>
                    </div>
                </div>
            `).join('');
        }
    } catch (err) {
        console.error("Failed to load AI recommendations:", err);
    }
}

// Add Schedule Modal
async function openAddScheduleModal() {
    const modal = document.getElementById('add-schedule-modal');
    if (!modal) return;

    // Populate origin and destination dropdowns with grouped optgroups
    try {
        const res = await fetch(`${API_BASE}/stations`);
        const data = await res.json();

        if (data.stations) {
            const originEl = document.getElementById('sched-origin');
            const destEl = document.getElementById('sched-destination');

            const lines = ['Purple Line', 'Green Line', 'Yellow Line'];
            let html = '';
            lines.forEach(line => {
                const lineStations = data.stations.filter(s => s.line === line);
                if (lineStations.length > 0) {
                    html += `<optgroup label="${line}">`;
                    lineStations.forEach(s => {
                        html += `<option value="${s.name}">${s.name}</option>`;
                    });
                    html += `</optgroup>`;
                }
            });

            if (originEl) originEl.innerHTML = html;
            if (destEl) destEl.innerHTML = html;
        }
    } catch (err) {
        console.error("Failed to load station options for schedule modal:", err);
    }

    modal.classList.add('active');
}

async function handleAddScheduleSubmit(e) {
    e.preventDefault();
    const trainCode = document.getElementById('sched-train-code').value.trim();
    const line = document.getElementById('sched-line').value;
    const origin = document.getElementById('sched-origin').value.trim();
    const destination = document.getElementById('sched-destination').value.trim();
    const departure = document.getElementById('sched-departure').value;
    const arrival = document.getElementById('sched-arrival').value;
    const frequency = parseInt(document.getElementById('sched-frequency').value);

    try {
        const res = await fetch(`${API_BASE}/schedules`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                train_code: trainCode,
                line: line,
                origin_station: origin,
                destination_station: destination,
                departure_time: departure,
                arrival_time: arrival,
                frequency_mins: frequency
            })
        });

        const data = await res.json();
        if (res.ok) {
            showToast("New train schedule added successfully!", "success");
            closeModal('add-schedule-modal');
            loadSchedulesPage();
        } else {
            showToast(data.error || "Failed to add schedule", "error");
        }
    } catch (err) {
        showToast("Error connecting to server.", "error");
    }
}

document.addEventListener('DOMContentLoaded', () => {
    if (document.getElementById('schedules-tbody')) {
        loadSchedulesPage();
    }
});

File: backend/routes/schedules.py
from flask import Blueprint, request, jsonify, session
from database import query_db, execute_db

schedules_bp = Blueprint('schedules', __name__, url_prefix='/api/schedules')

@schedules_bp.route('', methods=['GET'])
def get_schedules():
    line_filter = request.args.get('line')
    if line_filter:
        schedules = query_db('SELECT * FROM train_schedules WHERE line = ? ORDER BY departure_time ASC', [line_filter])
    else:
        schedules = query_db('SELECT * FROM train_schedules ORDER BY line ASC, departure_time ASC')
    return jsonify({'schedules': schedules})

@schedules_bp.route('', methods=['POST'])
def create_schedule():
    if session.get('role') != 'admin':
        return jsonify({'error': 'Unauthorized: Admin privilege required'}), 403

    data = request.get_json() or {}
    train_code = data.get('train_code')
    line = data.get('line')
    origin = data.get('origin_station')
    destination = data.get('destination_station')
    departure = data.get('departure_time')
    arrival = data.get('arrival_time')
    frequency = data.get('frequency_mins', 10)
    status = data.get('status', 'On Time')

    if not all([train_code, line, origin, destination, departure, arrival]):
        return jsonify({'error': 'Missing required schedule fields'}), 400

    try:
        new_id = execute_db('''
            INSERT INTO train_schedules (train_code, line, origin_station, destination_station, departure_time, arrival_time, frequency_mins, status)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?)
        ''', [train_code, line, origin, destination, departure, arrival, frequency, status])
        return jsonify({'message': 'Train schedule created successfully', 'id': new_id}), 201
    except Exception as e:
        return jsonify({'error': f'Failed to create schedule: {str(e)}'}), 400

@schedules_bp.route('/<int:schedule_id>', methods=['PUT'])
def update_schedule(schedule_id):
    if 'user_id' not in session:
        return jsonify({'error': 'Authentication required'}), 401

    data = request.get_json() or {}
    status = data.get('status')
    frequency = data.get('frequency_mins')

    if not status and frequency is None:
        return jsonify({'error': 'No update fields provided'}), 400

    existing = query_db('SELECT * FROM train_schedules WHERE id = ?', [schedule_id], one=True)
    if not existing:
        return jsonify({'error': 'Schedule record not found'}), 404

    new_status = status if status else existing['status']
    new_freq = frequency if frequency is not None else existing['frequency_mins']

    execute_db('''
        UPDATE train_schedules 
        SET status = ?, frequency_mins = ? 
        WHERE id = ?
    ''', [new_status, new_freq, schedule_id])

    return jsonify({'message': 'Schedule updated successfully', 'id': schedule_id})

@schedules_bp.route('/<int:schedule_id>', methods=['DELETE'])
def delete_schedule(schedule_id):
    if session.get('role') != 'admin':
        return jsonify({'error': 'Unauthorized: Admin privilege required'}), 403

    execute_db('DELETE FROM train_schedules WHERE id = ?', [schedule_id])
    return jsonify({'message': 'Schedule deleted successfully'})

@schedules_bp.route('/recommendations', methods=['GET'])
def get_recommendations():
    """
    AI-driven Train Frequency Recommendation Engine based on station crowd density.
    Computes optimum train headway for each metro line.
    """
    stations = query_db('SELECT * FROM stations')
    schedules = query_db('SELECT * FROM train_schedules')

    lines = list(set(s['line'] for s in stations))
    recommendations = []

    for line in lines:
        line_stations = [s for s in stations if s['line'] == line]
        if not line_stations:
            continue

        max_density = max(s['current_density'] for s in line_stations)
        busiest_station = max(line_stations, key=lambda s: s['current_density'])

        # Rule-based optimization logic driven by predicted/current density
        if max_density >= 80.0:
            rec_freq = 3  # High frequency peak dispatch
            action = 'URGENT: Increase Frequency'
            reason = f"Critical overcrowding detected at {busiest_station['name']} ({max_density}% capacity)."
        elif max_density >= 65.0:
            rec_freq = 5  # Peak frequency
            action = 'Increase Frequency'
            reason = f"High passenger volume at {busiest_station['name']} ({max_density}% capacity)."
        elif max_density >= 40.0:
            rec_freq = 8  # Normal frequency
            action = 'Maintain Regular Schedule'
            reason = f"Moderate traffic flow across {line}."
        else:
            rec_freq = 12  # Low frequency off-peak
            action = 'Reduce Frequency / Save Energy'
            reason = f"Low footfall across {line} (Peak density: {max_density}%)."

        current_line_schedules = [sch for sch in schedules if sch['line'] == line]
        avg_current_freq = sum(sch['frequency_mins'] for sch in current_line_schedules) / len(current_line_schedules) if current_line_schedules else 10

        recommendations.append({
            'line': line,
            'busiest_station': busiest_station['name'],
            'current_max_density': max_density,
            'current_avg_frequency': round(avg_current_freq, 1),
            'recommended_frequency': rec_freq,
            'action': action,
            'reason': reason
        })

    return jsonify({'recommendations': recommendations})
